### DISCUSSION TITLE
1. 네트워크 - 응용/표현/세션 계층

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#155] [질의 응답]
응용/표현/세션 계층 질의응답

**Comment by zini9188** [comment]
## Q. 응용 계층이란 무엇이고 어떤 역할을 하나요?

응용 계층은 OSI 7계층의 최상위 계층으로 인터페이스를 제공하며 사용자에게 최종적으로 보이는 계층이기도 합니다.

응용계층에는 STMP,  HTTP, FTP 등 여러가지 프로토콜이 존재합니다.

**Reply by zini9188** [reply]
## Q. HTTP 프로토콜에 대해 설명해주세요.

HTTP 프로토콜은 서버와 클라이언트간의 TCP/IP 프로토콜 위에서 데이터를 주고 받기 위한 프로토콜입니다. 

서버에 연결하여 요청 후, 응답을 받으면 연결을 끊는 비연결성 방식으로 작동됩니다. 연결을 끊어버리기 때문에 서버가 클라이언트의 이전 상태를 알 수 없어 무상태성의 특징을 가집니다.

**Reply by seongeun42** [reply]
## Q. 프로토콜이란 무엇인가요?

통신 프로토콜(=통신 규약)은 컴퓨터나 원거리 통신 장비 사이에서 메시지를 주고 받는 양식과 규칙의 체계를 말합니다.

**Comment by aeroej** [comment]
## Q. 표현 계층의 주요 기능은 무엇인가요?

표현 계층은 데이터의 변환, 암호화, 압축을 담당합니다. 
이 계층은 데이터 형식 간의 변환을 지원하여 데이터가 송신자와 수신자 사이에서 올바르게 전송될 수 있도록 합니다.

**Comment by seongeun42** [comment]
## Q. 세션 계층은 무엇인가요?

세션 계층은 장치 간의 통신을 관리하고 제어하는 역할을 합니다. 통신 세션을 설정, 유지, 종료, 에러 복구 등의 기능을 가지고 있습니다.

**Comment by watty-exclamationMark** [comment]
## Q. 이름 해석(Name resolution)이란 무엇인가요?
네트워크에서는 컴퓨터 or 네트워크에 붙여진 이름을 기반으로 IP주소를 알아냅니다.
이를 이름 해석이라고 부르고, 이름 해석을 위해 사용하는 것이 바로 DNS입니다.

**Reply by sikutisa** [reply]
## Q. DNS(Domain Name Service)에 대해 설명해주세요
* iP 주소 및 기타 데이터를 저장하고 이름별로 쿼리할 수 있게 해주는 계층형 분산 데이터베이스입니다.

**Reply by sikutisa** [reply]
## Q. (심화) DNS는 왜 계층형 분산 구조를 채택했다고 생각하시나요?
* 하나의 DNS를 사용시 DNS가 단일 장애점(single point of failure)이 됩니다.
* DNS에 대한 트래픽 부하가 과도해집니다.
* DNS와 물리적 거리가 먼 이용자 입장에서 속도가 너무 느립니다.
* (각주) TLD DNS, root DNS, authoritative DNS, local DNS와 iterated query, recursive query에 대해 찾아보면 좋다

**Comment by sikutisa** [comment]
## Q. 클라이언트-서버 구조(Client-Server Architecture)란 무엇인가요?
* 서버간의 통신을 서버-클라이언트로 분리시킨 구조입니다.
* 서버: 클라이언트의 요청에 따라 적절한 응답(리소스)를 전달해주는 곳으로 필요에 따라 서버는 데이터베이스에 요청을 보내고, 회신 받은 응답(리소스)을 활용합니다.
* 클라이언트: 서버로 요청을 보내고, 요청한 내용 응답(리소스)를 받아 사용합니다
* (각주) 위 구조는 2티어 구조이고, 데이터베이스가 들어가면 3티어 구조이다.

**Reply by sikutisa** [reply]
## Q. P2P 구조(P2P Architecture) 란 무엇인가요?
* 임의의 엔드포인트(호스트)들끼리 직접 통신하는 방식으로 어떤 호스트던간에 클라이언트의 역할을 할 수도, 서버의 역할을 할 수도 있습니다.
* 항상 기동하는(always-on) 서버가 없습니다.
* (각주) self scalability에 대해서도 찾아보자.

**Comment by sikutisa** [comment]
## Q. 소캣(Socket) 이란 무엇인가요?
* 어플리케이션 프로세스와 transport(전송) 계층 사이의, 통신을 주고 받는 일종의 통로입니다.
* (각주) 어플리케이션 계층과 전송 계층 사이가 아님을 주의

**Comment by sikutisa** [comment]
## Q. HTTP란 무엇인가요?
* HyperText Transfer Protocol의 약자로 HTML 문서와 같은 리소스들을 가져올 수 있도록 해주는 프로토콜입니다.
* 클라이언트는 브라우저를 통해서 서버에 요청을 보내고, 그 응답을 받아 웹 오브젝트를 화면에 표시합니다/
* 서버는 클라이언트의 요청에 대한 웹 오브젝트를 응답으로 전송합니다.
* (각주) HTTP 메시지 포맷에 대해 찾아보면 좋다

**Reply by sikutisa** [reply]
## Q. HTTP 연결의 종류
* non-persistent: TCP 연결 한 번에 최대 하나의 웹 오브젝트만 전송 가능합니다.
* persistent: TCP 연결 한 번에 여러 개의 객체를 전송할 수 있습니다.
* (각주) Rount Trip Time에 대해 찾아보면 좋다

**Reply by sikutisa** [reply]
## Q. HTTP 응답 코드 종류에 대해 설명해주세요
* 2XX: Success(성공)을 의미합니다.
* 3XX: Redirection(리다이렉션)을 의미합니다.
* 4XX: Client Error(클라이언트 에러)을 의미합니다.
* 5XX: Server Error(서버 에러)을 의미합니다.

**Comment by sikutisa** [comment]
## Q. 쿠키(Cookie) 란 무엇인가요?
* 웹사이트 접속시 접속자의 개인장치에 다운로드 되고 브라우저에 저장되는 작은 텍스트 파일입니다. 
* 웹사이트는 쿠키를 통해 접속자의 장치를 인식하고, 접속자의 설정과 과거 이용내역에 대한 일부 데이터를 저장합니다.

**Comment by sikutisa** [comment]
## Q. 웹 캐시(Web Caches) 또는 프록시 서버(Proxy Server)란 무엇인가요?
* 클라이언트가 자신을 통해서 다른 네트워크 서비스에 간접적으로 접속할 수 있게 해 주는 컴퓨터 시스템이나 응용 프로그램입니다.
* 원본 서버에 http 요청을 보내지 않고, 클라이언트의 요청에 응답을 해줌으로써 원본 서버의 부하를 줄여주기 위한 용도입니다.
